tests/test_single_topology_v3.py

```diff
@@ -383,3 +393,163 @@ def test_no_chiral_bond_restraints():
+    )
+
+    np.testing.assert_allclose(
+        f(0.0, k, lam),
+        f(k, 0.0, 1.0 - lam),
+        atol=1e-6,
+    )
+
+
+@given(nonzero_force_constants, st.lists(lambdas, min_size=3, max_size=3, unique=True).map(sorted))
+@seed(2022)
+def test_handle_ring_opening_closing_pin_to_end_states(k, lambdas):
+    lam, lambda_min, lambda_max = lambdas
+    assert handle_ring_opening_closing(linear_interpolation, 0.0, k, lam, lambda_min, lambda_max) == 0.0
+
+    lambda_min, lambda_max, lam = lambdas
+    assert handle_ring_opening_closing(linear_interpolation, 0.0, k, lam, lambda_min, lambda_max) == k
+
+
+@given(
+    nonzero_force_constants,
+    nonzero_force_constants,
+    nonzero_force_constants,
+    lambda_intervals,
+)
+@seed(2022)
+def test_interpolate_harmonic_force_constant(src_k, dst_k, k_min, lambda_interval):
+    lambda_min, lambda_max = lambda_interval
+
+    f = functools.partial(
+        interpolate_harmonic_force_constant,
+        k_min=k_min,
+        lambda_min=lambda_min,
+        lambda_max=lambda_max,
+    )
+
+    assert f(src_k, dst_k, 0.0) == src_k
+    assert f(src_k, dst_k, 1.0) == dst_k
+
+    lambdas = np.arange(0.01, 1.0, 0.01)
+
+    # all interpolated values >= k_min
+    np.testing.assert_array_less(1.0, f(src_k, dst_k, lambdas) / k_min + 1e-9)
+
+    def assert_nondecreasing(f):
+        y = f(lambdas)
+        np.testing.assert_array_less(y[:-1] / y[1:], 1.0 + 1e-9)
+
+    k1, k2 = sorted([src_k, dst_k])
+    assert_nondecreasing(lambda lam: f(k1, k2, lam))
+    assert_nondecreasing(lambda lam: f(k2, k1, 1.0 - lam))
+
+
+@given(
+    st.lists(
+        nonzero_force_constants,
+        min_size=2,
+        max_size=2,
+        unique=True,
+    ).filter(lambda ks: np.abs(ks[0] - ks[1]) / ks[1] > 1e-6)
+)
+@seed(2022)
+def test_interpolate_harmonic_force_constant_sublinear(ks):
+    src_k, dst_k = ks
+    lambdas = np.arange(0.01, 1.0, 0.01)
+    np.testing.assert_array_less(
+        interpolate_harmonic_force_constant(src_k, dst_k, lambdas, 1e-12, 0.0, 1.0)
+        / linear_interpolation(src_k, dst_k, lambdas),
+        1.0,
+    )
+
+
+def test_interpolate_harmonic_force_constant_jax_transformable():
+    _ = jax.jit(interpolate_harmonic_force_constant)(0.0, 1.0, 0.1, 1e-12, 0.0, 1.0)
+
+
+def test_cyclic_difference():
+    assert cyclic_difference(0, 0, 1) == 0
+    assert cyclic_difference(0, 1, 2) == 1  # arbitrary, positive by convention
+    assert cyclic_difference(0, 0, 3) == 0
+    assert cyclic_difference(0, 1, 3) == 1
+    assert cyclic_difference(0, 2, 3) == -1
+
+    # antisymmetric
+    assert cyclic_difference(0, 1, 3) == -cyclic_difference(1, 0, 3)
+    assert cyclic_difference(0, 2, 3) == -cyclic_difference(2, 0, 3)
+
+    # translation invariant
+    assert cyclic_difference(0, 1, 3) == cyclic_difference(-1, 0, 3)
+    assert cyclic_difference(0, 4, 8) == cyclic_difference(-2, 2, 8) == cyclic_difference(-4, 0, 8)
+
+    # jittable
+    _ = jax.jit(cyclic_difference)(0, 1, 1)
```

Comment:
tests are currently failing due to a ​`hypothesis​` exception related to inconsistent timings of `test_cyclic_difference_antisymmetric` -- possibly the first call is invoking the compiler https://gitlab.com/relaytx/physics-platform/timemachine/-/jobs/3092557213#L1431 . If so, possibly setting​`deadline=None`​ , or adding a `​jit(cyclic_difference)(inputs_with_expected_shapes_and_types)` ​statement prior to all tests would resolve.

Reply:
Just saw this and had arrived at a similar conclusion. I added a CI configuration profile setting `deadline=None` in [bbe57bc](https://github.com/proteneer/timemachine/pull/820/commits/bbe57bc7d013d6bfe387cd9b9327e6115bcd1546).